tls/s2n_crl.c

```diff
@@ -0,0 +1,85 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "s2n_crl.h"
+#include "tls/s2n_connection.h"
+
+static struct s2n_x509_crl* s2n_x509_crl_new(void) {
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_x509_crl)));
+
+    struct s2n_x509_crl *crl = (struct s2n_x509_crl*)(void*) mem.data;
+    crl->crl = NULL;
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return crl;
+}
+
+int s2n_x509_crl_from_pem(uint8_t *pem, size_t len, struct s2n_x509_crl **crl) {
+    POSIX_ENSURE_REF(crl);
+
+    struct s2n_blob pem_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&pem_blob, pem, len));
+
+    struct s2n_stuffer pem_stuffer = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&pem_stuffer, &pem_blob));
+    POSIX_GUARD(s2n_stuffer_skip_write(&pem_stuffer, pem_blob.size));
+
+    DEFER_CLEANUP(struct s2n_stuffer der_out_stuffer = {0}, s2n_stuffer_free);
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&der_out_stuffer, 2048));
```

Comment:
Where does this 2048 come from? Do we know the final size of the der output, or will we potentially need to resize this growable stuffer?

Reply:
It's coming from `s2n_x509_validator`, which does a similar pem to der conversion:
https://github.com/aws/s2n-tls/blob/af80f6b47a17bab71cfdc5377e3d1af7530ba540/tls/s2n_x509_validator.c#L89

Based on this, I think we do potentially need to resize it.

Reply:
But isn't that code called before the handshake? Like, when setting up the config? The performance requirements of setup code is much less of a concern than the performance of handshake code.

Can we estimate a reasonable upper bound for the output based on the input (while leaving it growable just in case)? So that in most cases we won't need to resize the stuffer repeatedly

Or is 2048 already the reasonable upper bound for most CRLs?

Reply:
Like resizing a stuffer isn't the end of the world and probably won't impact much, but we try to avoid it if possible

Reply:
Right, that makes sense. The size of the CRL really just depends on the number of serial numbers in it. Does it make sense to set it to the size of the pem? In my test CRL the pem was like 200 bytes larger than the der converted CRL, so this way the stuffer shouldn't have to grow: 87effb259c5c732cb1c66a8fe7fbb0fc9bdbc2b3

Reply:
I believe it will. I think the PEM/plaintext versions should always be bigger than the der ones.